### ISSUE TITLE
Support for Ledger Registrar Backers

### ISSUE THREAD

**Issue by SmithSamuelM** [decentralized-identity/keri#149]
Currently the only Backer type supported by KERI are witnesses that participate in a witness pool using the KAACE distributed consensus algorithm. Witnesses are identified with an ephemeral (non-transferable) identifier. The public key is in the identifier so no inception event or KEL is needed to verify signatures of a Witness. This is a performance optimization to avoid infinite regress of validating signatures o fwitnesses who have witnesses etc. When a witness' key is compromised, instead of rotating the witness keys, the witness identifier itself is rotated to a new ephemeral identifier.

However it has long been contemplated that KERI will support Backers that are registrars or oracles for shared ledgers such as Sovrin or Bitcoin etc.  Currently there is not mechanism to distinguish between the Backer types and a ledger registrar backer does not participate in a KAACE consensus pool with witness backers.  

The proposed mechanism here is that all non-witness backer types use non-ephemeral but non-transferable identifiers. A non-ephemeral but non-transferable identifier uses a transferable derivation code, thereby requiring an inception event but the nxt digest in its inception event is empty making it effectively non-transferable. But because an inception event is required we may use the seal list in the anchor field in the inception event to include a seal (digest) of all the configuration or meta data needed to know what ledger etc the  Backer is tied too. This incepting meta-date can be provided as an attachment to the inception event for that Backer's identifier. Using a non-ephemeral but effectively non-transferrable identifier may avoid the infinite regress in that a Backer may not have backers (other than the ledger itself that the backer is tied too).

This will require some code changes to detect non-witness Backers based on identifier type and inception event contents. This will be a different escrow path than witness backers. But the inception event allows us to securely designate any metadata needed to distinguish the Backer type and whatever else is needed to validate against that backer.

**Comment by pfeairheller** [comment]
If I understand this correctly, the use case is can be illustrated with the following example: 

An Indy Ledger (Sovrin, for example) wants to provide Ledger Backer support for KERI identifiers with their ledger.  So they would publish a non-ephemeral, non-transferable identifier inception event as a mechanism to publish the connection information needed to access their ledger.  That connection information would be an attachment to the inception event with a digest of the attachment contents in a seal in the inception event.

I understand the desire to avoid the infinite regress.  However, in the case of changing configuration information for the Ledger (adds/cuts of nodes, etc) wouldn't it be desirable to allow an event to commit to new configuration information without requiring everyone using this identifier as a witness to rotate to a new witness.  Since the backer protocol will already change requiring the ingestion of an inception event for a Ledger Registrar backer, wouldn't it make sense to allow for other events for these identifiers to change the configuration information?  This would also allow for key rotation if needed.

**Comment by SmithSamuelM** [comment]
I agree that one might want to have a TEL for the backer information or use a KEL for the backer keys itself.  The proposal is meant to be the minimally sufficient means to support backers.  We could allows variants that step up the security such as allowing multi-sig but still no rotation and then allowing rotation. However given that the original controller's KEL is the authoritative identifier for choosing backers. There is no reason from a security perspective to allow the backer to have rotatable keys.  Discovery of the backer's identifiers is provided by the Controller's KEL. So there is no need for those identifiers to have independent existence outside of the Controller's KEL.  

We spend so much effort supporting KELs that is easy to forget that the only reason one needs a KEL is for identifiers that are persistent and that means persistent across contexts. The Value of the identifier is that it has a reputation and a brand that needs to transcend or persist indefinitely. Otherwise we more simply just use ephemeral or effectively ephemeral identifiers that only exist within a given limited in time and space context.  So design of systems should always ask the question: does this identifier and its controlling key(s) need to persist outside of its immediate context? If not then we don't need an independent KEL or at least one that is transferable.    IN the case of Backers their value is solely a function of the context provided by the controller of the KEL to which they are a backer. The controller may rotate backer identifiers, therefore a backer identifier is effectively ephemeral not persistent. Thus they can have dedicated local to context identifiers.  If their keys get compromised the controller just rotates the identifiers. The same entity that controls the backer's key(s) could be  the same backer entity but each rotation changes the identifer/key set. No reason to independently maintain a transferable KEL.  Indeed that same backer entity good be a backer for multiple KELs but would use a dedicated specific identifier for each KEL.     

This approach also better manages privacy. For example if a KEL is meant to be private but wants backers for more security then the backer identifiers want to be private as well.  If a KEL is meant to be public, then the backer identifiers only become public because they are disclosed by that KEL. Having a separate KEL transferable key makes it harder to manage that privacy.   

To reiterate the 99% use case for KERI is that identifiers will be ephemeral, and if not ephemeral, i.e. persistent they will be privately persisting. It is only the 1% case where identifiers are public persistent. And when ephemeral the private vs. public should be a function of the privacy vs public of any persistent identifiers providing their context.

**Comment by SmithSamuelM** [comment]
Another way to think of this is that a single persistent identifier with its KEL provides a root-of-trust for an ecosystem of ephemeral or effectively ephemeral identifiers that all share that root-of-trust. Adding more complexity by adding other KELs other than trivial ones (i.e. inception only) is just that more complexity but does not improve the fundamental security of that -root-of-trust.

**Comment by SmithSamuelM** [comment]
@pfeairheller     

> wouldn't it be desirable to allow an event to commit to new configuration information without requiring everyone using this identifier as a witness to rotate to a new witness.

If we set up a TEL for configuration information then the signer on the configuration information could be  the Backer's signature.   This is essentially how the KERI-DHT works for the ip addresses of witnesses. The DHT stores the equivalent of an A record for each witness. The witness is responsible for signing and registering with the DHT (which is a type of TEL) a recored that links its witness identifier to its IP address.  

My point is that should that witnesses signature every become compromised. There is no reason that the DHT should ever need to keep track of KEL for that witness in order that the witness can rotate its keys. The controller of the KEL for which that witness is a witness may simply rotate to a new witness identifier with new keys and that new witness identifier then registers its IP address (could be the same one). The DHT only accepts A records for a witness if that witness is a valid witness for some KEL in the DHT. So when a controller rotates out a witness it also effectively revokes the A record of that witness in the DHT.

So in the case were we are not using ephemeral Backer identifiers like we are with witnesses but are using  effectively non-transferable identifiers that have only an inception event (we need some new terminology).  I think ephemeral non-transferable and non-ephemeral non-transferable might be better.  Then that inception event is for all intents and purposes the TEL for that Backer. Its a one shot TEL.   

So the real question is are there use cases where a backer would need to change its configuration without changing its identifier. Recall that the identifier when self addressing is strongly bound to the contents of its inception event. It is as secure as it will ever be.  If we change its configuration then how is that change secured.  We would need a TEL for that backer and a new event in the controller's KEL to anchor that TEL transaction that changed configuration. This would only make sense if the TEL anchors were in IXN interaction events, because if they were in a rotation event, then the controller could have more simply replaced the backer with a new one via the rotation instead of indirectly changing its configuration.  But now we have a case where a verifier of that KEL can't depend on the configuration of a given backer they have to lookup in the TEL. 

 So maybe there are use cases where configuration changes often enough that we don't want to be rotating to change backer configuration. But the use case we are taking about is Backers on Sovrin and the configuration for Sovrin should not change at all unless for example the access keys for the Backer to register events on the Sovrin ledger become compromised and in that case we are not simply changing Backer configuration but changing backers and rotation in the KEL is the secure way to protect recovery from key compromise, including Backer keys (which is how we do it for witness keys).  So for me the default should be backer configuration does not have a dedicated TEL unless it is something that is not key related like an IP address and then we let the DHT be the TEL.   

We could also have a default TEL for each KEL that is for non-key related backer information that acts in lieu of a DHT for discovery.  But we could just attach that information to an interaction event in the KEL that seals the attached information (this is how TrustFrame is planning to do it.)

The problem is discovery. How does a validator know that that information has changed?  So a default database TEL for metadata would be one way to solve it.  This is how everyone is using a DID:DOC its essentially  a database record of any configuration information that one wants to discover about a DID. Only its insecure because there is no provenance log. So a TEL based database has an provenance log so its secure by default and is the correct way to maintain a database of configuration information.


This is a long winded analysis. But the question is a good one because there are multiple ways to answer it. I am trying to delineate the options and trade-offs.

In summary:

Ledger Registrar Backers: The vital configuration is their access identifier on the ledger (bitcoin address). This is functionally a type  of ephemeral identifier (no provenance log for rotations). It is not a KERI ephemeral identifier because it does not use the KERI syntax for derivation codes and format. But functionally it is ephemeral. So it has the same security posture as a KERI witness.  Once that registrar creates a transaction on the ledger then that transaction is protected by the ledger. So no need for a watcher network (the ledger is both a witness and watcher network unto itself). Changing the ledger is not possible without creating a new access identifier (ethereum address for example) with new private key. This is a type of key rotation and so we want to protect that by rotating the Backer to a new Backer from a bitcoin backer to an ethereum backer and the most secure way to do that is to do a KEL rotation.  When I first worked through the KERI events, I finalized on only allowing witness changes with a rotation (establishment) event , not an interaction (non-establishment) event. The reason is that I could easily come up with exploits where compromised signing keys could change the security posture of a recovery rotation event by first changing the witness set to one the compromiser controlled and then when the controller attempted to recover via a rotation the recovery was not protected by witnesses the controller controlled anymore.  We would have the same issue with non-witness backers. TrustFrame has a related problem for NFTs because they did not make a commitment to restrict the backer set in their provenance log. 

The policy for KERI is that if a Backer's keys change then that change requires a Rotation in the backed KEL. But if the Backer's identifier is a self-addressing identifier derived from the public keys (either in the inception event or as an ephemeral non-transferable identifier) then its identifier would also change. So this would create a new backer identifier not a change in configuration of that backer.

Now if all I want to do is change the IP address of the Backer then that is a discovery problem and the keri-dht already solves that.  But in lieu of a KERI dht, each KEL could have a default TEL that provides other non-key based "backer config" including witness backer config not merely ledger registrar backer config such as IP addresses. This TEL becomes a secure replacement for a did:doc and the did method for KERI could be extended to provide this default config TEL state in the did:doc.

**Comment by SmithSamuelM** [comment]
To summarize again.

The policy is that Backers can only endorse something previously committed to via a seal in the KEL for which the backer is a backer.

This brings full circle the meaning of endorse. (BTW endorser is not a word in the dictionary, only the verb endorse is).

Backers may only endorse a prior commitment made by the controller. This means that a validator may drop any signed statement by a backer that has not already been anchored on the KEL. 

This applies to a witnesses in a witness pool are a single ledger registrar. IN the case of a witness pool the independent keys of each witness provide security via a threshold structure. In the case of a ledger registrar the ledger access keys of the registrar enable it to cross anchor the controller's commitment in its KEL on a ledger. This provides duplicity protection via the ledger instead of a watcher network. The first registration on the Ledger wins because ledgers are totally ordering. If a ledger registrar's access key(s) become compromised then the controller rotates to a new registrar with new key(s). A validator then may ignore any new entries made by that old registrar after it has been rotated out because they won't have corresponding prior commitments in the KEL. Only the new registrar will.

The registrar's KERI identifier is derived from an inception event where the public key is a KERI formatted version of that registrars access public key. This binding tightly the access identifier on the ledger to the Keri identifier used to designated that registrar as a backer.    

For example bitcoin: Bitcoin address is a fingerprint of a EcDSA_secp_256_k1 public key. We put that public key in the inception event for that backer registrar and derive the resultant KERI identifier. So all the KERI attachments etc use that identifier but any validator can derive the bitcoin corresponding bitcoin address and confirm which transactions on the bitcoin ledger come from the bitcoin version of that same public key.

For example Sovrin: Sovrin Nym is a fingerprint of a Ed25519 public key. We put that public key in the inception event and the rest is that same as above.

**Comment by pfeairheller** [comment]
> Now if all I want to do is change the IP address of the Backer then that is a discovery problem and the keri-dht already solves that. But in lieu of a KERI dht, each KEL could have a default TEL that provides other non-key based "backer config" including witness backer config not merely ledger registrar backer config such as IP addresses. This TEL becomes a secure replacement for a did:doc and the did method for KERI could be extended to provide this default config TEL state in the did:doc.

This is the essence of what I was asking and describes a solution approach pretty well...

But this:
> The witness stills signs that publication as a spam DDoS protection but now the controller doesn't care about monitoring the IP addresses of its witnesses that have been published to the DHT. The DHT won't update a witness A record without first validating that the update has been anchored in the KEL.

And this:
> There is no explicit TEL for backer configuration metadata. All the TEL data is attached to the event that anchors it in the original KEL and when a validator replays a KEL it extracts the attachments and store them locally.

Describe the solution that I think you just summarized.   Changing the DHT to help address this is a win, in my opinion.

**Comment by SmithSamuelM** [comment]
Yes I just had to reason through all the options to get clarity in my own mind.
